subprojects/docs/src/samples/user-input/publishing-credentials/groovy/build.gradle

```diff
@@ -0,0 +1,53 @@
+plugins {
+    id 'java-library'
+    id 'maven-publish'
+}
+
+version = '1.0.2'
+group = 'com.example'
+
+// tag::publication[]
+def MAVEN_USERNAME_PROPERTY = 'mavenUser'
+def MAVEN_PASSWORD_PROPERTY = 'mavenPassword'
+def mavenUser = providers.gradleProperty(MAVEN_USERNAME_PROPERTY)
+def mavenPassword = providers.gradleProperty(MAVEN_PASSWORD_PROPERTY)
+
+publishing {
+    publications {
+        library(MavenPublication) {
+            from components.java
+        }
+    }
+    repositories {
+        maven {
+            name = 'mySecure'
+            url = uri(com.example.MavenRepositoryStub.address())
+            credentials {
+                username = mavenUser.orNull
+                password = mavenPassword.orNull
+            }
+        }
+    }
+}
+// end::publication[]
+
+// tag::validation[]
+tasks.named('publishLibraryPublicationToMySecureRepository') {
+    doFirst {
+        if (!mavenUser.present || !mavenPassword.present) {
+            throw new GradleException(String.format("Publishing requires '%s' and '%s' properties",
```

Comment:
It would be nice if providers would either have a validation method that would fail with a descriptive error message and/or be describable in some way, maybe just toString saying what type of provider it is and its name. Now toString yields `provider provider(?)`. More helpful would be `Gradle property 'X' provider`.
Like here - I want to validate that a provider is present and have a descriptive error message about what expectation is not satisfied and why.
For Gradle properties provider, it could say that it could not find the required property named X, looked in the following places for it, for more details link to documentation chapter about Gradle properties.

Reply:
Another option - instead of the doFirst block, we could say:
```
gradle.taskGraph.whenReady { taskGraph ->
    if (taskGraph.allTasks.any { it.name == 'publishLibraryPublicationToMySecureRepository' }) {
        if (!mavenUser.present || !mavenPassword.present) {
            throw new GradleException(String.format("Publishing requires '%s' and '%s' properties",
                    MAVEN_USERNAME_PROPERTY, MAVEN_PASSWORD_PROPERTY))
        }
    }
}
```
This way the validation can be performed before any tasks are run and thus fail fast.